src/main/java/hudson/remoting/forward/CopyThread.java

```diff
@@ -16,10 +16,23 @@
     private final InputStream in;
     private final OutputStream out;
 
-    public CopyThread(String threadName, InputStream in, OutputStream out) {
+    public CopyThread(String threadName, InputStream in, OutputStream out, Runnable termination) {
+        this(threadName, in, out, termination, 0);
+    }
+
+    private CopyThread(String threadName, InputStream in, OutputStream out, Runnable termination, int previousTries) {
         super(threadName);
         this.in = in;
         this.out = out;
+        setUncaughtExceptionHandler((t, e) -> {
+            if (previousTries < 5) {
+                LOGGER.log(Level.FINE, "Uncaught exception in CopyThread " + t + ", retrying copy", e);
+                new CopyThread(threadName, in, out, termination, previousTries + 1).start();
```

Comment:
Maybe it should be a WARNING even for allowed retries. Even if the copythread recovers, it may be a dataloss if the input stream has been read without writing. Or maybe the thread should be able to send cause to the termination handler? It would be able to process it in a custom way then. It would require some extra API, but it seems to be reasonable.